cirq/circuits/circuit_operation_test.py

```diff
@@ -167,31 +196,70 @@ def test_with_params():
         _ = cirq.resolve_parameters(op_base, cirq.ParamResolver(param_dict))
 
 
-def test_repetition():
-    a, b = cirq.LineQubit.range(2)
-    # This circuit has a modulus of 8.
-    circuit = cirq.FrozenCircuit(cirq.H(a), cirq.CX(a, b))
-    op_base = cirq.CircuitOperation(circuit)
-    assert op_base.repeat(1) is op_base
+def cartesian_product_of_string_lists(list1: List[str], list2: List[str]):
+    return [f'{first}-{second}' for first in list1 for second in list2]
 
-    op_with_reps = op_base.repeat(-5)
-    assert op_with_reps.repetitions == -5
-    assert op_base ** -5 == op_with_reps
 
-    with pytest.raises(TypeError):
-        _ = op_base.repeat(1.3)
-
-
-def test_repeat_measurement_fails():
+@pytest.mark.parametrize('add_measurements', [True, False])
+@pytest.mark.parametrize('use_default_ids_for_initial_rep', [True, False])
+def test_repeat(add_measurements, use_default_ids_for_initial_rep):
     a, b = cirq.LineQubit.range(2)
-    circuit = cirq.FrozenCircuit(
-        cirq.X(a),
-        cirq.measure(b, key='mb'),
-        cirq.measure(a, key='ma'),
-    )
-    op = cirq.CircuitOperation(circuit)
-    with pytest.raises(NotImplementedError):
-        _ = op.repeat(3)
+    circuit = cirq.Circuit(cirq.H(a), cirq.CX(a, b))
+    if add_measurements:
+        circuit.append([cirq.measure(b, key='mb'), cirq.measure(a, key='ma')])
+    op_base = cirq.CircuitOperation(circuit.freeze())
+    assert op_base.repeat(1) is op_base
+    assert op_base.repeat(1, ['0']) != op_base
+    assert op_base.repeat(1, ['0']) == op_base.repeat(repetition_ids=['0'])
+    assert op_base.repeat(1, ['0']) == op_base.with_repetition_ids(['0'])
+
+    initial_repetitions = -3
+    if add_measurements:
+        with pytest.raises(ValueError, match='circuit is not invertible'):
+            _ = op_base.repeat(initial_repetitions)
+        initial_repetitions = abs(initial_repetitions)
+
+    final_repetitions = 2 * initial_repetitions
```

Comment:
This can be moved closer to where it's used, for readability.

Reply:
Done